lib/logger/logger.nit

```diff
@@ -0,0 +1,402 @@
+		var format = formatter or else default_formatter
+		if format == null then
+			return add_raw(level, message)
+		end
+		return add_raw(level, format.format(level, message))
+	end
+
+	# Output a message with `level` severity without formatting it
+	#
+	# Only output messages with `level` severity greater than of equal to `self.level`.
+	#
+	# ~~~
+	# var stderr = new StringWriter
+	# var logger = new Logger(warn_level, stderr, null)
+	#
+	# # This message will be displayed:
+	# assert logger.warn("This is a warning.")
+	# assert stderr.to_s.trim.split("\n").last == "This is a warning."
+	#
+	# # This message will not:
+	# assert not logger.info("This is some info.")
+	# assert stderr.to_s.trim.split("\n").last == "This is a warning."
+	# ~~~
+	fun add_raw(level: Int, message: Writable): Bool do
+		if level < self.level then return false
+		out.write(message.write_to_string)
+		out.write("\n")
+		return true
+	end
+
+	# Output a message with `unknown_level` severity
+	#
+	# Unkown severity messages are always outputted.
+	fun unknown(message: String, formatter: nullable Formatter): Bool do
+		return add(unknown_level, message, formatter)
+	end
+
+	# Output a message with `fatal_level` severity
+	fun fatal(message: String, formatter: nullable Formatter): Bool do
+		return add(fatal_level, message, formatter)
+	end
+
+	# Output a message with `error_level` severity
+	fun error(message: String, formatter: nullable Formatter): Bool do
+		return add(error_level, message, formatter)
+	end
+
+	# Output a message with `warn_level` severity
+	fun warn(message: String, formatter: nullable Formatter): Bool do
+		return add(warn_level, message, formatter)
+	end
+
+	# Output a message with `info_level` severity
+	fun info(message: String, formatter: nullable Formatter): Bool do
+		return add(info_level, message, formatter)
+	end
+
+	# Output a message with `debug` severity
+	fun debug(message: String, formatter: nullable Formatter): Bool do
+		return add(debug_level, message, formatter)
+	end
+end
+
+# Log messages to a file
+#
+# ~~~
+# var log_file = "my_file.log"
+# var logger = new FileLogger(warn_level, log_file, append = false)
+# logger.error("An error")
+# logger.info("Some info")
+# logger.close
+# assert log_file.to_path.read_all == "An error\n"
+#
+# logger = new FileLogger(warn_level, log_file, append = true)
+# logger.error("Another error")
+# logger.close
+# assert log_file.to_path.read_all == "An error\nAnother error\n"
+#
+# log_file.to_path.delete
+# ~~~
+class FileLogger
+	super Logger
+	autoinit level, file, append, default_formatter
+
+	redef type OUT: FileWriter
+
+	# File where messages will be written
+	var file: String
+
+	# Append messages to `file`
+	#
+	# If `append` is `false`, the `file` will be overwritten.
+	var append: Bool = true is optional
+
+	init do
+		var old = null
+		if append then
+			old = file.to_path.read_all
+		end
+		out = new FileWriter.open(file)
```

Comment:
Oh, it seems we do not have an `append` mode on files in Nit.

Reply:
I was surprised too.

Reply:
There is still an issue about the `stream` lib tho: https://github.com/nitlang/nit/issues/466